pytorch_lightning/trainer/distrib_data_parallel.py

```diff
@@ -528,7 +528,7 @@ def ddp_train(self, process_idx, q, model, is_master=False, proc_offset=0):
             if is_master:
                 # source of truth is cuda for gpu idx
                 gpus = os.environ['CUDA_VISIBLE_DEVICES'].split(',')
-                gpu_idx = int(gpus[self.local_rank])
+                gpu_idx = self.local_rank
```

Comment:
this won’t work... because if you have access to gpus 4,5,6,7 and you request “2,3” you’re asking for “6,7”

Reply:
> if you have access to gpus 4,5,6,7 and you request “2,3” you’re asking for “6,7”

In my view, having PL run your mode on GPUs 6,7 is the correct behavior in this case.

Reply:
This fixes another problem with ddp. If `gpus=3` and `CUDA_VISIBLE_DEVICES=4,5,6,7`, ddp will run only two jobs on GPUs 5,6, and the job on GPU4 won't work.

Reply:
i agree but here's what happens:

gpus available: 0, 1, 2, 3, 4, 5
index:                0, 1, 2, 3, 4, 5
gpu[2] = 2

when you set CUDA_VISIBLE_DEVICES your numbering changes
CUDA_VISIBLE_DEVICES='2, 4,5'
now your indexes               0, 1, 2

So once you set visible devices the mapping changes:
gpus[0] = 2
gpus[2] = 5

Reply:
can you share code that breaks so i can reproduce and verify? your fix might fix this problem but it's likely to break other DDP settings